dandi/metadata.py

```diff
@@ -83,3 +88,169 @@ def get_metadata(path):
     meta["nd_types"] = get_neurodata_types(path)
 
     return meta
+
+
+def parse_age(age):
+    """
+    Convert a human-friendly duration string into an ISO 8601 duration
+
+    Parameters
+    ----------
+    age : str
+
+    Returns
+    -------
+    str
+    """
+    m = re.fullmatch(r"(\d+)\s*(y(ear)?|m(onth)?|w(eek)?|d(ay)?)s?", age, flags=re.I)
+    if m:
+        qty = int(m.group(1))
+        unit = m.group(2)[0].upper()
+        return f"P{qty}{unit}"
+    else:
+        raise ValueError(age)
+
+
+def extract_age(metadata):
+    try:
+        dob = ensure_datetime(metadata["date_of_birth"])
+        start = ensure_datetime(metadata["session_start_time"])
+    except (KeyError, TypeError, ValueError):
+        try:
+            duration = parse_age(metadata["age"])
+        except (KeyError, TypeError, ValueError):
+            return ...
+    else:
+        if start < dob:
+            raise ValueError("session_start_time precedes date_of_birth")
+        duration = timedelta2duration(start - dob)
+    return models.PropertyValue(value=duration, unitText="Years from birth")
+
+
+def timedelta2duration(delta):
+    """
+    Convert a datetime.timedelta to ISO 8601 duration format
+
+    Parameters
+    ----------
+    delta : datetime.timedelta
+
+    Returns
+    -------
+    str
+    """
+    s = "P"
+    if delta.days:
+        s += f"{delta.days}D"
+    if delta.seconds or delta.microseconds:
+        sec = delta.seconds
+        if delta.microseconds:
+            # Don't add when microseconds is 0, so that sec will be an int then
+            sec += delta.microseconds / 1000000
+        s += f"T{sec}S"
+    if s == "P":
+        s += "0D"
+    return s
+
+
+def extract_sex(metadata):
+    if metadata.get("sex") is not None:
+        return models.SexType(identifier="sex", name=metadata["sex"])
```

Comment:
for each of these extractors we need to create a search tool. one that maps the existing metadata value to some known entity. i'll try to create a quick first pass at this.  but i think we are close now. we need to do a similar metadata map for the existing dandiset metadata schema as well.